src/rgw/rgw_kms.cc

```diff
@@ -40,6 +35,228 @@ static void concat_url(std::string &url, std::string path) {
+  {
+    bufferlist secret_bl;
+    int res;
+    string vault_token = "";
+    if (RGW_SSE_KMS_VAULT_AUTH_TOKEN == cct->_conf->rgw_crypt_vault_auth){
+      ldout(cct, 0) << "Loading Vault Token from filesystem" << dendl;
+      res = load_token_from_file(&vault_token);
+      if (res < 0){
+        return res;
+      }
+    }
+
+    std::string secret_url = cct->_conf->rgw_crypt_vault_addr;
+    if (secret_url.empty()) {
+      ldout(cct, 0) << "ERROR: Vault address not set in rgw_crypt_vault_addr" << dendl;
+      return -EINVAL;
+    }
+
+    concat_url(secret_url, cct->_conf->rgw_crypt_vault_prefix);
+    concat_url(secret_url, std::string(key_id));
+
+    RGWHTTPTransceiver secret_req(cct, "GET", secret_url, &secret_bl);
+
+    if (!vault_token.empty()){
+      secret_req.append_header("X-Vault-Token", vault_token);
+      vault_token.replace(0, vault_token.length(), vault_token.length(), '\000');
+    }
+
+    string vault_namespace = cct->_conf->rgw_crypt_vault_namespace;
+    if (!vault_namespace.empty()){
+      ldout(cct, 20) << "Vault Namespace: " << vault_namespace << dendl;
+      secret_req.append_header("X-Vault-Namespace", vault_namespace);
+    }
+
+    res = secret_req.process(null_yield);
+    if (res < 0) {
+      ldout(cct, 0) << "ERROR: Request to Vault failed with error " << res << dendl;
+      return res;
+    }
+
+    if (secret_req.get_http_status() ==
+        RGWHTTPTransceiver::HTTP_STATUS_UNAUTHORIZED) {
+      ldout(cct, 0) << "ERROR: Vault request failed authorization" << dendl;
+      return -EACCES;
+    }
+
+    ldout(cct, 20) << "Request to Vault returned " << res << " and HTTP status "
+      << secret_req.get_http_status() << dendl;
+
+    ldout(cct, 20) << "Parse response into JSON Object" << dendl;
+
+    if (!parser->parse(secret_bl.c_str(), secret_bl.length())) {
+      ldout(cct, 0) << "ERROR: Failed to parse JSON response from Vault" << dendl;
+      return -EINVAL;
+    }
+    secret_bl.zero();
+
+    return res;
+  }
+
+  int decode_secret(JSONObj* json_obj, std::string& actual_key){
+    std::string secret;
+    try {
+      secret = from_base64(json_obj->get_data());
+    } catch (std::exception&) {
+      ldout(cct, 0) << "ERROR: Failed to base64 decode key retrieved from Vault" << dendl;
+      return -EINVAL;
+    }
+
+    actual_key.assign(secret.c_str(), secret.length());
+    secret.replace(0, secret.length(), secret.length(), '\000');
+    return 0;
+  }
+
+public:
+
+  VaultSecretEngine(CephContext *cct) {
+    this->cct = cct;
+  }
+
+  virtual ~VaultSecretEngine(){}
+};
+
+
+class TransitSecretEngine: public VaultSecretEngine {
+
+private:
+  int get_key_version(boost::string_view key_id, string& version)
+  {
+    size_t pos = 0;
+
+    pos = key_id.rfind("/");
+    if (pos != std::string::npos){
+      boost::string_view token = key_id.substr(pos+1, key_id.length()-pos);
+      if (!token.empty() && token.find_first_not_of("0123456789") == std::string::npos){
```

Comment:
since key_id is a boost::string_view, we should use boost::string_view::npos instead of std::string::npos